packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs

```diff
@@ -1,42 +1,70 @@
 
 use crate::{
-    knowledge::PersistedLink,
+    knowledge::{Link, LinkQueryPath, PersistedLink},
+    ontology::LinkTypeQueryPath,
     store::{
         crud,
-        postgres::context::PostgresContext,
-        query::{Expression, ExpressionError, Literal},
+        postgres::query::{Distinctness, Ordering, SelectCompiler},
+        query::Filter,
         AsClient, PostgresStore, QueryError,
     },
 };
 
 #[async_trait]
 impl<C: AsClient> crud::Read<PersistedLink> for PostgresStore<C> {
-    type Query<'q> = Expression;
+    type Query<'q> = Filter<'q, Link>;
 
     async fn read<'f: 'q, 'q>(
         &self,
-        query: &'f Self::Query<'q>,
+        filter: &'f Self::Query<'q>,
     ) -> Result<Vec<PersistedLink>, QueryError> {
-        // TODO: We need to work around collecting all records before filtering
-        //   related: https://app.asana.com/0/1202805690238892/1202923536131158/f
-        stream::iter(self.read_all_links().await?.collect::<Vec<_>>().await)
-            .try_filter_map(|record| async move {
-                if let Literal::Bool(result) = query
-                    .evaluate(&record, self)
-                    .await
-                    .change_context(QueryError)?
-                {
-                    Ok(result.then(|| PersistedLink::from(record)))
-                } else {
-                    bail!(
-                        Report::new(ExpressionError)
-                            .attach_printable("does not result in a boolean value")
-                            .change_context(QueryError)
-                    );
-                }
+        let mut compiler = SelectCompiler::new();
+        compiler.add_selection_path(
+            &LinkQueryPath::Type(LinkTypeQueryPath::VersionedUri),
+            Distinctness::Destinct,
+            None,
+        );
+        compiler.add_selection_path(&LinkQueryPath::Source(None), Distinctness::Destinct, None);
+        compiler.add_selection_path(&LinkQueryPath::Target(None), Distinctness::Destinct, None);
+        compiler.add_selection_path(
+            &LinkQueryPath::Index,
+            Distinctness::Destinct,
+            Some(Ordering::Ascending),
+        );
+        compiler.add_selection_path(&LinkQueryPath::OwnedById, Distinctness::Indestinct, None);
+        compiler.add_selection_path(&LinkQueryPath::CreatedById, Distinctness::Indestinct, None);
+        compiler.add_filter(filter);
+
+        let (statement, parameters) = compiler.compile();
+
+        self.as_client()
+            .query_raw(&statement, parameters.iter().copied())
+            .await
+            .into_report()
+            .change_context(QueryError)?
+            .map(|row| row.into_report().change_context(QueryError))
+            .and_then(|row| async move {
+                let link_type_id = VersionedUri::from_str(row.get(0))
+                    .into_report()
+                    .change_context(QueryError)?;
+                let source_entity_id = row.get(1);
```

Comment:
Are the orders of the results always stable? As in do they always follow the order of `add_selection_path`?

Reply:
Yes, they are stable 🙂